plugins/ndpi/ndpi.c

```diff
@@ -0,0 +1,624 @@
+                        ndpi_detection_giveup(threadctx->ndpi, flowctx->ndpi_flow, &proto_guessed);
+                flowctx->detection_completed = 1;
+            }
+        }
+
+        if (flowctx->detection_completed) {
+            SCLogNotice("Detected protocol: %s | app protocol: %s | category: %s",
+                    ndpi_get_proto_name(
+                            threadctx->ndpi, flowctx->detected_l7_protocol.proto.master_protocol),
+                    ndpi_get_proto_name(
+                            threadctx->ndpi, flowctx->detected_l7_protocol.proto.app_protocol),
+                    ndpi_category_get_name(
+                            threadctx->ndpi, flowctx->detected_l7_protocol.category));
+        }
+    }
+}
+
+static void OnFlowFinish(ThreadVars *tv, Flow *f, void *_data)
+{
+    /* Nothing to do here, the storage API has taken care of cleaning
+     * up storage, just here for example purposes. */
+    SCLogDebug("Flow %p is now finished", f);
+}
+
+static void OnThreadInit(ThreadVars *tv, void *_data)
+{
+    struct NdpiThreadContext *context = SCCalloc(1, sizeof(*context));
+    if (context == NULL) {
+        FatalError("Failed to allocate nDPI thread context");
+    }
+    context->ndpi = ndpi_init_detection_module(NULL);
+    if (context->ndpi == NULL) {
+        FatalError("Failed to initialize nDPI detection module");
+    }
+    NDPI_PROTOCOL_BITMASK protos;
+    NDPI_BITMASK_SET_ALL(protos);
+    ndpi_set_protocol_detection_bitmask2(context->ndpi, &protos);
+    ndpi_finalize_initialization(context->ndpi);
+    ThreadSetStorageById(tv, thread_storage_id, context);
+}
+
+static int DetectnDPIProtocolPacketMatch(
+        DetectEngineThreadCtx *det_ctx, Packet *p, const Signature *s, const SigMatchCtx *ctx)
+{
+    const Flow *f = p->flow;
+    struct NdpiFlowContext *flowctx = FlowGetStorageById(f, flow_storage_id);
+    const DetectnDPIProtocolData *data = (const DetectnDPIProtocolData *)ctx;
+    bool r;
+
+    SCEnter();
+
+    /* if the sig is PD-only we only match when PD packet flags are set */
+    /*
+    if (s->type == SIG_TYPE_PDONLY &&
+            (p->flags & (PKT_PROTO_DETECT_TS_DONE | PKT_PROTO_DETECT_TC_DONE)) == 0) {
+        SCLogDebug("packet %"PRIu64": flags not set", p->pcap_cnt);
+        SCReturnInt(0);
+    }
+    */
+
+    if (!flowctx->detection_completed) {
+        SCLogDebug("packet %" PRIu64 ": ndpi protocol not yet detected", p->pcap_cnt);
+        SCReturnInt(0);
+    }
+
+    if (f == NULL) {
+        SCLogDebug("packet %" PRIu64 ": no flow", p->pcap_cnt);
+        SCReturnInt(0);
+    }
+
+    r = ndpi_is_proto_equals(flowctx->detected_l7_protocol.proto, data->l7_protocol, false);
+    r = r ^ data->negated;
+
+    if (r) {
+#if 0
+        SCLogDebug("ndpi protocol match on protocol = %u.%u (match %u)",
+                flowctx->detected_l7_protocol.app_protocol, flowctx->detected_l7_protocol.master_protocol,
+                data->l7_protocol);
+#endif
```

Comment:
@cardigliano Can you check this? It fails to build.

Reply:
Replace flowctx->detected_l7_protocol with flowctx->detected_l7_protocol.proto (fixed in our branch)

Reply:
The 3rd argument is bad as well:

https://github.com/ntop/suricata/blob/ndpi-integration-v1/plugins/ndpi/ndpi.c#L210

Should it be `data->l7_protocol.master_protocol` or `app_protocol`?